game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -61,71 +97,95 @@ public void retreatUnits(final IDelegateBridge bridge) {
+      return new RetreaterAirAmphibious(battleState);
     }
+    return null;
   }
 
-  @Value(staticConstructor = "of")
-  private static class RetreatData {
-    RetreatType retreatType;
-    Collection<Territory> retreatSites;
-  }
+  private void retreat(final IDelegateBridge bridge, final Retreater retreater) {
+    final Collection<Unit> retreatUnits = retreater.getRetreatUnits();
+    final Collection<Territory> possibleRetreatSites =
+        retreater.getPossibleRetreatSites(retreatUnits);
+    final String text = retreater.getQueryText();
+
+    bridge.getDisplayChannelBroadcaster().gotoBattleStep(battleState.getBattleId(), getName());
+    final Territory retreatTo =
+        battleActions.queryRetreatTerritory(
+            battleState, bridge, battleState.getAttacker(), possibleRetreatSites, text);
+    if (retreatTo == null) {
+      return;
+    }
+    SoundUtils.playRetreatType(
+        battleState.getAttacker(), retreatUnits, retreater.getRetreatType(), bridge);
 
-  private boolean canAttackerRetreat() {
-    return RetreatChecks.canAttackerRetreat(
-        battleState.getUnits(BattleState.Side.DEFENSE),
-        battleState.getGameData(),
-        battleState::getAttackerRetreatTerritories,
-        battleState.isAmphibious());
-  }
+    final CompositeChange change = new CompositeChange();
 
-  private boolean canAttackerRetreatSeaPlanes() {
-    return battleState.getBattleSite().isWater()
-        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+    change.add(retreater.extraRetreatChange(retreatTo, retreatUnits));
+
+    final Map<Retreater.RetreatLocation, Collection<Unit>> retreatingUnitMap =
+        retreater.splitRetreatUnits(retreatUnits);
+
+    retreatSameTerritoryUnits(
+        bridge, retreatingUnitMap.get(Retreater.RetreatLocation.SAME_TERRITORY));
+    change.add(
+        retreatOtherTerritoryUnits(
+            bridge, retreatTo, retreatingUnitMap.get(Retreater.RetreatLocation.OTHER_TERRITORY)));
+
+    bridge.addChange(change);
+
+    if (battleState.getUnits(BattleState.Side.OFFENSE).isEmpty()) {
+      battleActions.endBattle(IBattle.WhoWon.DEFENDER, bridge);
+    } else {
+      bridge.getDisplayChannelBroadcaster().notifyRetreat(battleState.getBattleId(), retreatUnits);
+    }
+
+    bridge
+        .getDisplayChannelBroadcaster()
+        .notifyRetreat(
+            battleState.getAttacker().getName() + retreater.getShortBroadcastSuffix(),
+            battleState.getAttacker().getName() + retreater.getLongBroadcastSuffix(retreatTo),
+            getName(),
+            battleState.getAttacker());
   }
 
-  private boolean canAttackerRetreatPartialAmphib() {
-    if (!Properties.getPartialAmphibiousRetreat(battleState.getGameData())) {
-      return false;
+  private void retreatSameTerritoryUnits(
+      final IDelegateBridge bridge, final Collection<Unit> retreatingUnits) {
+    if (retreatingUnits != null && !retreatingUnits.isEmpty()) {
```

Comment:
IMO the control flow would be better at the caller level. This method is a no-op otherwise an is just a void call from the caller. It provides some abstraction but a maintainer needs to jump around in the code to see that this is a potential no-op and when it would be a no-op. It also looks like the null check is dead-code.

I suspect some details are a bit hidden that probably could be exposed to the caller that would make the code easier to follow. For example, instead of this:

```
  @Override
  public Map<RetreatLocation, Collection<Unit>> splitRetreatUnits(
      final Collection<Unit> retreatUnits) {
    return Map.of(RetreatLocation.SAME_TERRITORY, retreatUnits);
  }
```

Perhaps the API should be:

```
Collection<Unit> splitRetreatUnits(
     RetreatLocation,
      final Collection<Unit> retreatUnits)
```

In this way it'd be more obvious that if we are calling a non-nullable method, then we'd always have a key value for 'retreatingUnits' and one would then be very comfortable removing the null check.